### PR TITLE
Plotting – again

### PR DESCRIPTION
**Pull request by bertsky**
After several attempts by @Shreeshrii to share her excellent plotting scripts, each of which was unfortunately thwarted by bad circumstances (other big changes occurring at the same time), here comes a plotting facility again.

I based this on the [ocrddata branch](https://github.com/Shreeshrii/tesstrain/tree/ocrddata) of her fork, cherry-picking only the two relevant changesets, resolving conflicts and then refactoring to make this better fit our makefileization.

Usage is simply `make plot`, which will only work after `make training`. (I could also make this dependency explicit, but that would cause `make plot` to start the training if it did not happen already for that combination of variables.)

The output files will be created in `$OUTPUT_DIR/$MODEL_NAME.plot_log.png`, e.g.
herrnhut-kurrent tess finetuned-htrbin plot_log

and `$OUTPUT_DIR/$MODEL_NAME.plot_cer.png`, e.g.
herrnhut-kurrent tess finetuned-htrbin plot_cer

All intermediate files (except for the lstmeval log files generated under `$OUTPUT_DIR/eval/*.log` because they are valuable in their own right) are marked as such and therefore removed by make.

Perhaps we should discuss how both plots could be combined into a single one (which is probably what @Shreeshrii tried to do already) – I can see that there's a problem by the granularity these data points are recorded (training iterations for validation during lstmtraining vs. learning iterations for validation afterwards via external lstmeval). But IIUC we have everything it takes to be able to combine them (twin y plot with synced x axes)...

### CODE REVIEW

**Comment by bertsky** [comment]
With the last commit I did re-instante @Shreeshrii's LOG_FILE variable. 

The big pro is that thus you can opt in to plotting even older logs, e.g. 

    LOG_FILE=nohup.out make plot

**Comment by zdenop** [comment]
I just make quick test on openSUSE (15.5) and here are a few suggestions:

-  it would be nice to have a short example of how to make example plot on example data: 
```sh
git clone https://github.com/tesseract-ocr/tesstrain
cd tesstrain
mkdir data
unzip ocrd-testset.zip -d data/ocrd-ground-truth
...
# install needed requirements
...
nohup make training MODEL_NAME=ocrd START_MODEL=frk TESSDATA=~/tessdata_best MAX_ITERATIONS=10000 > plot/TESSTRAIN.LOG &
make plot MODEL_NAME=ocrd
```

-  I removed python2 from openSUSE and I got this error:
```sh
python plot_cer.py data/ocrd ocrd data/ocrd/ocrd.iteration.tsv data/ocrd/ocrd.checkpoint.tsv data/ocrd/ocrd.eval.tsv data/ocrd/ocrd.sub.tsv data/ocrd/ocrd.lstmeval.tsv
/bin/bash: python: command not found
```
What about using `PY_CMD` (as rest of Makefile)?

- When I run manually `python3 plot_cer.py data/ocrd ocrd data/ocrd/ocrd.iteration.tsv data/ocrd/ocrd.checkpoint.tsv data/ocrd/ocrd.eval.tsv data/ocrd/ocrd.sub.tsv data/ocrd/ocrd.lstmeval.tsv` I got error:
```sh
Traceback (most recent call last):
  File "/home/podobny/Projekty/tesstrain/plot_cer.py", line 6, in <module>
    import matplotlib
ModuleNotFoundError: No module named 'matplotlib'
```
It would be good to mentioned that user should install matplotlib and pandas (`pip3 install matplotlib pandas`) before running `make plot`.

**Comment by stweil** [comment]
> It would be good to mention that user should install matplotlib and pandas (`pip3 install matplotlib pandas`) before running `make plot`.

Both are mentioned in `requirements.txt`, so running `pip3 install -r requirements.txt` is sufficient.

**Comment by bertsky** [comment]
@zdenop very good points – thanks! 
I'll address these in a few follow-up commits.

File: Makefile
Comment (stweil) [inline Makefile:399]:
```suggestion
$(TSV_CHECKPOINT): $(LOG_FILE)
```

File: Makefile
Comment (stweil) [inline Makefile:408]:
```suggestion
$(TSV_EVAL): $(LOG_FILE)
```

File: README.md
Comment (stweil) [inline README.md:118]: I would prefer a different target name _evaluation_.

File: README.md
Comment (stweil) [inline README.md:114]: That additional information is an internal detail (which might change in the future). I suggest to keep the old text.

**Review by stweil — CHANGES_REQUESTED** [review]
@bertsky, some of the commits could already be applied to the main branch. Would it be okay if I cherry-pick them? You (or I) would have to rebase your plotting branch after that. I think we can improve plotting faster by getting it integrated like that.

File: Makefile
Comment (bertsky) [reply Makefile:113]: Please make use of suggestions next time – this hickups my workflow (I must rebase what I already have).

File: Makefile
Comment (bertsky) [reply Makefile:388]: well spotted!

File: README.md
Comment (bertsky) [reply README.md:114]: The descriptions of `traineddata` and (new) `lstmeval` already refer to it, though.

File: .gitignore
Comment (bertsky) [reply .gitignore:13]: I did not touch this myself. Sure, that would be better. Lots of other lines here I like to prune.

**Comment by bertsky** [comment]
> some of the commits could already be applied to the main branch. Would it be okay if I cherry-pick them? You (or I) would have to rebase your plotting branch after that. I think we can improve plotting faster by getting it integrated like that.

@stweil why the hurry all of a sudden? This does not bode well with my workflow, and I'm already in too many places at a time...

**Comment by bertsky** [comment]
All done.
@zdenop do you want me to add some target for doing the pip install requirements, too? 

Also, how about adding the result of the plotting for the ocrd-testset training into the repo and showing it in the readme?

**Comment by zdenop** [comment]
Ok for me. After merging, I will try to check it on Windows... (I have some ideas for improvements already)

File: Makefile
Comment (stweil) [inline Makefile:425]:
```suggestion
	$(PY_CMD) plot_log.py $@ $(MODEL_NAME) $(TSV_100_ITERATIONS) $(TSV_CHECKPOINT) $(TSV_EVAL) $(TSV_SUB)
```

File: Makefile
Comment (stweil) [inline Makefile:427]:
```suggestion
	$(PY_CMD) plot_cer.py $@ $(MODEL_NAME) $(TSV_100_ITERATIONS) $(TSV_CHECKPOINT) $(TSV_EVAL) $(TSV_SUB) $(TSV_LSTMEVAL)
```

**Review by stweil — CHANGES_REQUESTED** [review]
I suggest to omit the `$(MODEL_NAME)_` prefixes.

File: plot_cer.py
Comment (stweil) [inline plot_cer.py:52]:
```suggestion
    tmin = z.iloc[np.argmin(y)]
    xmin = x.iloc[np.argmin(y)]
```

File: plot_log.py
Comment (stweil) [inline plot_log.py:51]:
```suggestion
          xmin = x.iloc[np.argmin(y)]
```

File: plot_log.py
Comment (stweil) [inline plot_log.py:56]:
```suggestion
          tmin = z.iloc[np.argmin(y)]
          xmin = x.iloc[np.argmin(y)]
```

**Review by stweil — CHANGES_REQUESTED** [review]
These changes fix some runtime warnings:
```
tesstrain/plot_cer.py:51: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  tmin = z[np.argmin(y)]
tesstrain/plot_cer.py:52: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  xmin = x[np.argmin(y)]
tesstrain/plot_log.py:55: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  tmin = z[np.argmin(y)]
tesstrain/plot_log.py:56: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  xmin = x[np.argmin(y)]
```

File: Makefile
Comment (bertsky) [reply Makefile:424]: No, that would mean the generated image file name does not contain the model name anymore. Once you copy it to another directory, the info is gone and it might clash with other plots.

File: Makefile
Comment (bertsky) [reply Makefile:426]: same here.

File: plot_cer.py
Comment (bertsky) [reply plot_cer.py:52]: care to explain this?

File: Makefile
Comment (bertsky) [reply Makefile]: I did go the same route eventually (just not via your suggestions, sry)

File: plot_cer.py
Comment (bertsky) [reply plot_cer.py:50]: fair enough – got to know – thx!

**Comment by bertsky** [comment]
So?

**Comment by zdenop** [comment]
@stweil  I see no warning mentioned by you:

> tesstrain/plot_log.py:55: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]` tmin = z[np.argmin(y)])

I have pandas 2.2.0. What is your pandas version?

@bertsky : please reformat the Python scripts with [blue](https://pypi.org/project/blue/) (or `black`) there is plenty of formal formatting issues (missing spaces).

and [ruff](https://github.com/astral-sh/ruff) complains about this:
```sh
ruff check plot_cer.py
plot_cer.py:43:1: E741 Ambiguous variable name: `l`
Found 1 error.
```

**Comment by bertsky** [comment]
In light of https://github.com/tesseract-ocr/tesseract/issues/3763#issuecomment-2028441608 I tend to prefer changing from `fast` to `best` models for evaluation.

**Comment by bertsky** [comment]
@zdenop 
> and [ruff](https://github.com/astral-sh/ruff) complains about this:
> 
> ```shell
> ruff check plot_cer.py
> plot_cer.py:43:1: E741 Ambiguous variable name: `l`
> Found 1 error.
> ```

I fail to see the ambiguity. The [manual says](https://docs.astral.sh/ruff/rules/ambiguous-variable-name/) they think l can be easily confused with 1. I really don't care for such subjective stances.

> please reformat the Python scripts with [blue](https://pypi.org/project/blue/) (or `black`) there is plenty of formal formatting issues (missing spaces).

Not my code originally, so I don't care. But I also don't believe in code stylers. Since @stweil seems to be an avid user of these, I don't think my help is needed in that respect.

I'll resolve the conflict on the readme arising from your concurrent edits in the plotting section, then I'll be done here, I think.

**Review by zdenop — APPROVED** [review]

File: .gitignore
Comment (stweil) [inline .gitignore:1]:
```suggestion
/data/
```

File: Makefile
Comment (stweil) [inline Makefile:131]:
```suggestion
	@echo "    training         Start training"
```

Comment (stweil) [reply Makefile:131]: I'd prefer a less technical description here and avoid details which might change any time.

File: Makefile
Comment (stweil) [inline Makefile:135]:
```suggestion
	@echo "    evaluation       Evaluate intermediate models on evaluation dataset"
```

Comment (stweil) [reply Makefile:135]: Use help text with less technical details

File: Makefile
Comment (stweil) [inline Makefile:161]:
```suggestion
	@echo "    NET_SPEC           Network specification (in VGSL), only used for training without START_MODEL. Default: $(NET_SPEC)"
```

File: Makefile
Comment (stweil) [inline Makefile:167]:
```suggestion
	@echo "    LOG_FILE           File copy of the training protocol (also used for plotting). Default: $(LOG_FILE)"
```

File: README.md
Comment (stweil) [inline README.md:113]:
```suggestion
    training         Start training
```

File: README.md
Comment (stweil) [inline README.md:210]:
```suggestion
Still, the target will depends on the `LOG_FILE` captured during training (it will just not trigger training itself).
Besides the log file, the trained models (for each checkpoint) are also evaluated on the eval dataset.
The latter is also available as an independent target `evaluation`:
```

**Review by stweil — CHANGES_REQUESTED** [review]
@bertsky, I still have several change requests for the help text in Makefile. Those changes then must be made in README.md as well.

File: Makefile
Comment (bertsky) [reply Makefile:131]: Like I [said above](https://github.com/tesseract-ocr/tesstrain/pull/377#discussion_r1516591680), the description of `traineddata` already mentions .checkpoint files, too. And so does `evaluation`. So as a matter of clarity and consistency, it should be mentioned here as well.

Since this entire repo is just a small makefile wrapper around the actual training tools, which are only properly documented in tessdoc _along these technical details_, I prefer keeping it the way it is.

File: Makefile
Comment (bertsky) [reply Makefile:135]: see above.

File: Makefile
Comment (bertsky) [reply Makefile:167]: I don't like your formulation. LOG_FILE is not a protocol of the training procedure (like Tensorboard), it's just the output of the training recipe.

File: README.md
Comment (bertsky) [reply README.md]: `will depends`?

I don't like passive voice here, but pushed an improved paragraph.

File: README.md
Comment (bertsky) [reply README.md:217]: I like `depict` better than `show` here, and `has changed` better than `is changed`.

File: Makefile
Comment (bertsky) [reply Makefile:161]: I found it more helpful if the relevant terms `fine-tune` and `scratch` were already mentioned in their respective variable's short help string. (Since `START_MODEL` already mentions the former, `NET_SPEC` now mentions the latter.) Also, `scratch` already precludes a start model by definition.

File: Makefile
Comment (bertsky) [reply Makefile:135]: No, I _don't agree_, hence my comment above.

Comment (bertsky) [reply Makefile:135]: > Isn't your help text even wrong? I though that lstmeval is running on .traineddata files, not on .checkpoint files.

No, it's running on .traineddata extracted from .checkpoint files. The help texts only mention the former when relevant to the user (`traineddata` target).

File: Makefile
Comment (stweil) [reply Makefile:167]: If you prefer "log" instead of "protocol", that would be fine for me, too. The output of the training is much more than the printed messages: the most important output is the new model file.

File: README.md
Comment (stweil) [reply README.md:217]: Deepl and I disagree.

File: Makefile
Comment (bertsky) [reply Makefile:167]: A LOG_FILE is a log file and not a result file.

File: README.md
Comment (bertsky) [reply README.md:217]: Deepl does not know anything about tesstrain, and I am pretty confident in my assessment of what the English language allows me to express.

Besides, IMO it should not be your role as a maintainer to micromanage individual contributions to the letter. This is becoming extremely tedious – I thought this addition is long overdue and would be welcomed.

File: Makefile
Comment (stweil) [reply Makefile:167]: "Log file of training process (also used as input for plotting)" would also be fine for me.

**Pull request merged**
